extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ObserverValidationProcessor.java

```diff
@@ -0,0 +1,63 @@
+package io.quarkus.arc.deployment;
+
+import java.util.Collection;
+
+import javax.inject.Inject;
+
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.AnnotationValue;
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.IndexView;
+import org.jboss.logging.Logger;
+
+import io.quarkus.arc.processor.BeanDeploymentValidator;
+import io.quarkus.arc.processor.DotNames;
+import io.quarkus.arc.processor.ObserverInfo;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.ApplicationArchivesBuildItem;
+
+/**
+ * Validates observer methods from application classes.
+ * If an observer listening for {@code @Initialized(ApplicationScoped.class)} is found, it logs a warning.
+ */
+public class ObserverValidationProcessor {
+
+    private static final Logger LOGGER = Logger.getLogger(ObserverValidationProcessor.class.getName());
+
+    @Inject
+    BuildProducer<BeanDeploymentValidatorBuildItem> validators;
+
+    @BuildStep
+    public void validateApplicationObserver(ApplicationArchivesBuildItem applicationArchivesBuildItem) {
+        // an index of all application classes
+        IndexView applicationClassesIndex = applicationArchivesBuildItem.getRootArchive().getIndex();
+
+        validators.produce(new BeanDeploymentValidatorBuildItem(new BeanDeploymentValidator() {
+
+            @Override
+            public void validate(ValidationContext context) {
+                Collection<ObserverInfo> allObservers = context.get(Key.OBSERVERS);
+                // do the validation for each observer that can be found within application classes
+                for (ObserverInfo observer : allObservers) {
+                    DotName declaringBeanDotName = observer.getDeclaringBean().getBeanClass();
+                    AnnotationInstance annotationInstance = observer.getObserverMethod().annotation(DotNames.INITIALIZED);
```

Comment:
Just a little detail - `observer.getObserverMethod().annotation(DotNames.INITIALIZED)` will return a non-null value even if the annotation is accidentally declared on the method...

Reply:
Hmm, I don't think there is a way to query for just method param annotations.
Therefore, to avoid situations where you accidentally declare it on method and on method param, we should probably get all annotations and iterate over them?

Reply:
Yes, there is no such method in the API. And yes, you'll need to filter the annotation instances by target. Something like https://github.com/quarkusio/quarkus/blob/master/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Annotations.java#L54-L63.

Reply:
BTW feel free to add a utility method to the `Annotations` class.

Reply:
Hm, pushed changes before I read these comments, I'll add the util method too.